File: GroupDocs.Annotation.VisualBasic/GroupDocs.AnnotationDataStorage/JsonFile.vb
Imports Newtonsoft.Json
Imports System.IO

Public Class JsonFile(Of T As New)
#Region "Fields"
    Protected ReadOnly _syncRoot As New Object()

    Private _filePath As String
    Private _data As T
#End Region

    Public Sub New(filePath As String)
        If [String].IsNullOrWhiteSpace(filePath) Then
            Throw New ArgumentNullException("filePath")
        End If

        _filePath = filePath
    End Sub

    Protected Overridable Sub Serialize()
        If _data Is Nothing Then
            Return
        End If

        SyncLock _syncRoot
            If _data IsNot Nothing Then
                Try
                    Using stream = File.OpenWrite(_filePath)
                        Using writer = New StreamWriter(stream)
                            Using jwriter As JsonWriter = New JsonTextWriter(writer) With { _
                                .Formatting = Formatting.Indented _
                            }
                                Dim serializer As New JsonSerializer()
                                serializer.Serialize(jwriter, _data)
                            End Using
                        End Using
                    End Using
                Catch e As Exception
                    Throw New Exception("Failed to serialize an object to file: '{0}'.", e)
                End Try
            End If
        End SyncLock
    End Sub

    Protected Overridable Sub Deserialize()
        SyncLock _syncRoot
            Try
                If Not File.Exists(_filePath) Then
                    Dim fileStream = File.Create(_filePath)
                    fileStream.Close()
                    _data = New T()
                    Return
                End If

                Using stream = File.OpenRead(_filePath)
                    Using reader = New StreamReader(stream)
                        Using jreader As JsonReader = New JsonTextReader(reader)
                            Dim serializer As New JsonSerializer()
                            _data = serializer.Deserialize(Of T)(jreader)
                        End Using
                    End Using
                End Using
            Catch e As Exception
                Throw New Exception("Failed to deserialize an object from file: '{0}'.", e)
            End Try

            If _data Is Nothing Then
                _data = New T()
            End If
        End SyncLock
    End Sub

    Protected ReadOnly Property Data() As T
        Get
            SyncLock _syncRoot
                If _data Is Nothing Then
                    Deserialize()
                End If

                Return _data
            End SyncLock
        End Get
    End Property
End Class

File: Examples/GroupDocs.Annotation.CSharp/GroupDocs.AnnotationDataStorage/JsonFile.cs
using System;
using System.IO;
using Newtonsoft.Json;

namespace GroupDocs.Data.Json
{

    public class JsonFile<T>
        where T : new()
    {
        #region Fields
        protected readonly object _syncRoot = new object();

        private string _filePath;
        private T _data;
        #endregion Fields

        public JsonFile(string filePath)
        {
            if (String.IsNullOrWhiteSpace(filePath))
            {
                throw new ArgumentNullException("filePath");
            }

            _filePath = filePath;
        }

        protected virtual void Serialize()
        {
            if (_data == null)
            {
                return;
            }

            lock (_syncRoot)
            {
                if (_data != null)
                {
                    try
                    {
                        using (var stream = File.OpenWrite(_filePath))
                        using (var writer = new StreamWriter(stream))
                        using (JsonWriter jwriter = new JsonTextWriter(writer) {Formatting = Formatting.Indented})
                        {
                            JsonSerializer serializer = new JsonSerializer();
                            serializer.Serialize(jwriter, _data);
                        }
                    }
                    catch (Exception e)
                    {
                        throw new Exception("Failed to serialize an object to file: '{0}'.", e);
                    }
                }
            }
        }

        protected virtual void Deserialize()
        {
            lock (_syncRoot)
            {
                try
                {
                    if (!File.Exists(_filePath))
                    {
                        var fileStream = File.Create(_filePath);
                        fileStream.Close();
                        _data = new T();
                        return;
                    }

                    using (var stream = File.OpenRead(_filePath))
                    using (var reader = new StreamReader(stream))
                    using (JsonReader jreader = new JsonTextReader(reader))
                    {
                        JsonSerializer serializer = new JsonSerializer();
                        _data = serializer.Deserialize<T>(jreader);
                    }
                }
                catch (Exception e)
                {
                    throw new Exception("Failed to deserialize an object from file: '{0}'.", e);
                }

                if(_data == null)
                {
                    _data = new T();
                }
            }
        }

        protected T Data
        {
            get
            {
                lock (_syncRoot)
                {
                    if (_data == null)
                    {
                        Deserialize();
                    }

                    return _data;
                }
            }
        }
    }
}
